service/worker/diagnostics/invariants/timeout.go

```diff
@@ -93,3 +107,95 @@ func (t *timeout) Check(context.Context) ([]InvariantCheckResult, error) {
 	}
 	return result, nil
 }
+
+func (t *timeout) RootCause(ctx context.Context, issues []InvariantCheckResult) ([]InvariantRootCauseResult, error) {
+	result := make([]InvariantRootCauseResult, 0)
+	for _, issue := range issues {
+		pollerStatus, err := t.checkTasklist(ctx, issue)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, pollerStatus)
+
+		if issue.InvariantType == TimeoutTypeActivity.String() {
+			heartbeatStatus, err := checkHeartbeatStatus(issue)
+			if err != nil {
+				return nil, err
+			}
+			result = append(result, heartbeatStatus)
+		}
+	}
+	return result, nil
+}
+
+func (t *timeout) checkTasklist(ctx context.Context, issue InvariantCheckResult) (InvariantRootCauseResult, error) {
+	var taskList *types.TaskList
+	switch issue.InvariantType {
+	case TimeoutTypeExecution.String():
+		var metadata ExecutionTimeoutMetadata
+		err := json.Unmarshal(issue.Metadata, &metadata)
+		if err != nil {
+			return InvariantRootCauseResult{}, err
+		}
+		taskList = metadata.Tasklist
+	case TimeoutTypeActivity.String():
+		var metadata ActivityTimeoutMetadata
+		err := json.Unmarshal(issue.Metadata, &metadata)
+		if err != nil {
+			return InvariantRootCauseResult{}, err
+		}
+		taskList = metadata.Tasklist
+	}
+	if taskList == nil {
+		return InvariantRootCauseResult{}, fmt.Errorf("tasklist not set")
+	}
+
+	frontendClient := t.clientBean.GetFrontendClient()
+	resp, err := frontendClient.DescribeTaskList(ctx, &types.DescribeTaskListRequest{
+		Domain:   t.domain,
+		TaskList: taskList,
+	})
+	if err != nil {
+		return InvariantRootCauseResult{}, err
+	}
+
+	tasklistBacklog := resp.GetTaskListStatus().GetBacklogCountHint()
+	if len(resp.GetPollers()) == 0 {
+		return InvariantRootCauseResult{
+			RootCause: RootCauseTypeMissingPollers.String(),
+			Metadata:  taskListBacklogInBytes(tasklistBacklog),
+		}, nil
+	}
+	return InvariantRootCauseResult{
+		RootCause: RootCauseTypePollersStatus.String(),
+		Metadata:  taskListBacklogInBytes(tasklistBacklog),
+	}, nil
+
+}
+
+func checkHeartbeatStatus(issue InvariantCheckResult) (InvariantRootCauseResult, error) {
+	var metadata ActivityTimeoutMetadata
+	err := json.Unmarshal(issue.Metadata, &metadata)
+	if err != nil {
+		return InvariantRootCauseResult{}, err
+	}
+
+	if metadata.HeartBeatTimeout == 0 {
+		return InvariantRootCauseResult{
+			RootCause: RootCauseTypeHeartBeatingNotEnabled.String(),
+			Metadata:  []byte(metadata.TimeElapsed.String()),
+		}, nil
```

Comment:
This suppose to hit a lot - lots of activity calls don't use heartbeats.
Are you taking into account it is not as crucial as not the next if-statement?

Reply:
there is no prioritisation done here. we simply list potential root causes cos we cant quite predict what suits the usecase. but calling out heart beating not enabled is one of the many observations we return when an activity times out.